### PR TITLE
Fastify Type Providers

### PR DESCRIPTION
**Pull request by sinclairzx81**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

This PR is a follow on from issue #3391. It provides a working type plugin system to enable third party inference libraries to interface Fastify's type definitions. The implementation primarily focuses on setting up the necessary types to enable such plugins, but also implements the Type Provider mechanism which was mentioned on #3391. 

## Type Providers

Type Providers are mapping interfaces that can be extended by library authors to enable static type inference for specific kinds of data (such as JSON schema). A type provider interface is passed to Fastify via a new `.withTypeProvider<Provider>()` function. This function returns the same Fastify instance; but remapped to use the mapping implemented for the provider.

The Type Provider (essentially an HKT) can be thought of as a simple input / output mapping function that accepts the input of one type (i.e. schemas) and maps it to another. The current implementation is setup such that `Body`, `Headers`, `Params` and `Querystring` pipe through to this mapping function. This mechanism enables libraries to provide auxiliary type inference for Fastify without Fastify needing to take these libraries as dependencies.

The following demonstrates Type Provider usage by integrating both Json-Schema-To-Ts and TypeBox inference libraries.

> Note that the `TypeBoxProvider` and `JsonSchemaToTsProvider` may be offered as installable plugins.

### Json-Schema-to-Ts

```typescript
import Fastify, { FastifyTypeProvider } from 'fastify'

import { FromSchema, JSONSchema } from 'json-schema-to-ts'

// ---------------------------------------------------------
// Type Provider
// ---------------------------------------------------------

interface JsonSchemaToTsProvider extends FastifyTypeProvider {

  output: this['input'] extends JSONSchema ? FromSchema<this['input']> : never
}

// ---------------------------------------------------------
// Example
// ---------------------------------------------------------

const fastify = Fastify().withTypeProvider<JsonSchemaToTsProvider>()

fastify.get('/', {
  schema: {
    body: {
      type: 'object',
      properties: {
        x: { type: 'string' },
        y: { type: 'number' },
        z: { type: 'boolean' }
      },
      required: ['x', 'y', 'z']
    } as const
  }
}, (req, reply) => {

  // type Body = { x: string, y: number, z: boolean }

  const { x, y, z } = req.body
})
```

### TypeBox

```typescript
import Fastify, { FastifyTypeProvider } from 'fastify'

import { Type, Static, TSchema } from '@sinclair/typebox'

// ---------------------------------------------------------
// Type Provider
// ---------------------------------------------------------

interface TypeBoxProvider extends FastifyTypeProvider {

  output: this["input"] extends TSchema ? Static<this["input"]> : never
}

// ---------------------------------------------------------
// Example
// ---------------------------------------------------------

const fastify = Fastify().withTypeProvider<TypeBoxProvider>()

fastify.get('/', {
  schema: {
    body: Type.Object({
      x: Type.String(),
      y: Type.Number(),
      z: Type.Boolean()
    })
  }
}, (req, reply) => {

  // type Body = { x: string, y: number, z: boolean }

  const { x, y, z } = req.body
})
```

## Additional Work

This implementation only goes as far as implementing Type Providers for Fastify requests only. However there is potential to extend this functionality to Fastify responses as well. A supplemental project is offered with this PR which implements both parameter inference from URL as well as strict type checking on responses for Fastify. This project is offered for reference.

Project: https://github.com/sinclairzx81/fastify-typebox

TypeScript Demo Available [Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAMQIYGcbAGYE8A0cDecAKlmAKZwC+cGUEIcA5BqutgLQylkBGEAHowCwAKFEBjCADs0NVpixwAvInnYAFAEpRolmgUA6SGnWMA9AFcUZKCjMAuKzYCSAE0Z5CouD7gpxABZkIEj2BN6+kXyuWGEk5AYA8jwAVmTiMOr4EZG5cPxx3AYAchYgPDZaODl5vrHERaXlldoitVSa1W21UGQokDJkYdnd7T4ATAAMk4UJyWkZWTVjvr0oFgA2MLNkJWUVUFrLtZSt7ZQ1FyKUeOq9AI4WfTB4vWAbWJrKAHzho3A1MxmODsUFg8EQyFQ6HsQHAgBKZEezxQcJBMIxmMhOn+QLgXHIcAACkgoEgQChlAQ4E4oG4wmgoMApABzKg4yKSGTwQi0txUKkPJ5oIyk8mosS44EEigAIQgMSphAKcCk+xseHqauaUHZks50lkys1ApUQueBmiihWPg5vjxWMdWLRiPewD6aKdXuhdp8eJlcER-UNFBUhCmM2pa0221V6t11CuNTeHwM1ikriycGjWzCAEZph0bbk8TY6FAwlIIH4YEgYFY4JJXBR+ulMO7XL7s2R3lg07X6yh1AAWaaaNNkDNZnOx4fjIuRUtQcthUeTGt1htNihV+DNjDMsid-WrHuptCbocR8fpzOEGdhRhzxgLuBLlfd9ZbfHcODASm7nGOpdimfYXoO6jXhOU73n0MZhPwcAANRwNapx+sCEAANZhAAhKIpxAA)

### CODE REVIEW

File: docs/Type-Providers.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    const { foo, bar } = request.query // type safe!
+})
+```
+
+### TypeBox
+
+The following sets up a TypeBox Type Provider
+
+```bash
+$ npm install fastify-type-provider-typebox --save
```
[the remark this answers is not in the document]
Comment (sinclairzx81) [reply docs/Type-Providers.md:56]: Yes, I'd expect `fastify-type-provider-json-schema-to-ts` and `fastify-type-provider-typebox` would both belong as packages of the Fastify org. Each would take a direct dependency on `json-schema-to-ts` and `@sinclair/typebox` respectively as well as Fastify V4, likely as a peer dependency. 

Each provider package would contain a single module that exports an implementation of `FastifyTypeProvider`. The contents of these modules are as follows.

### fastify-type-provider-json-schema-to-ts
```typescript
import { FastifyTypeProvider } from 'fastify'

import { JSONSchema, FromSchema } from 'json-schema-to-ts'

export interface JsonSchemaToTsProvider extends FastifyTypeProvider {

  output: this['input'] extends JSONSchema ? FromSchema<this['input']> : never
}
```

### fastify-type-provider-typebox

```typescript
import { FastifyTypeProvider } from 'fastify'

import { Static, TSchema } from '@sinclair/typebox'

export * from '@sinclair/typebox'

export interface TypeBoxTypeProvider extends FastifyTypeProvider {

  output: this["input"] extends TSchema ? Static<this["input"]> : never
}
```
These packages are yet to be setup.

File: docs/TypeScript.md [hunk 2]
```diff
@@ -332,7 +332,7 @@ const todo = {
     done: { type: 'boolean' },
```
Comment (RafaelGSS) [inline docs/TypeScript.md:332]: Can you add somewhere an link to the `Type-Providers` document? I think that's valuable for the Typescript section.

File: docs/Type-Providers.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+### Providers
+
+Type Providers are offered as additional packages you will need to install into your project. Each provider uses a different inference library under the hood; allowing you to select the library most appropriate for your needs. Type Provider packages follow a `fastify-type-provider-{provider-name}` naming convention. 
+
+The following inference packages are supported:
+
+- `json-schema-to-ts` - [github](https://github.com/ThomasAribart/json-schema-to-ts)
+- `typebox` - [github](https://github.com/sinclairzx81/typebox)
+
+### Json Schema to Js
```
Comment (RafaelGSS) [inline docs/Type-Providers.md:16]:
```suggestion
### Json Schema to Ts
```

**Review by mcollina — APPROVED** [review]
lgtm

File: test/types/schema.test-d.ts [hunk 4]
```diff
@@ -35,33 +35,6 @@ expectAssignable<FastifyInstance>(server.get(
   () => { }
 ))
 
-expectAssignable<FastifyInstance>(server.get<RouteGenericInterface, ContextConfigDefault, { validate:(data: any) => any }>(
```
Comment (RafaelGSS) [inline test/types/schema.test-d.ts:38]: Why those tests were removed?

File: test/types/type-provider.test-d.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
+import fastify, { FastifyTypeProvider } from '../../fastify'
```
Comment (RafaelGSS) [inline test/types/type-provider.test-d.ts:1]: Can you add some assertions to other handlers? I mean:

```ts
preHandler: (req, res) => {
  expectType<X>(req.body)
}
```
```

File: types/typeProvider.d.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+export interface FastifyTypeProvider {
```
Comment (RafaelGSS) [inline types/typeProvider.d.ts:1]: Can you rename this file to `type-provider.d.ts`? just to follow the pattern in this folder.

**Review by RafaelGSS — COMMENTED** [review]
Few comments, huuuge work! Thanks.

File: test/types/schema.test-d.ts (hunk 4 above)
Comment (sinclairzx81) [reply test/types/schema.test-d.ts:38]: These were removed as per @mcollina 's request. But happy to re-add them.

File: test/types/type-provider.test-d.ts (hunk 5 above)
Comment (sinclairzx81) [reply test/types/type-provider.test-d.ts:1]: Have added an additional test for each request hook.

File: test/types/schema.test-d.ts (hunk 4 above)
Comment (sinclairzx81) [reply test/types/schema.test-d.ts:38]: @RafaelGSS Let me know if you would like me to re-add these tests.

File: docs/TypeScript.md (hunk 2 above)
Comment (sinclairzx81) [reply docs/TypeScript.md:332]: The Type Provider documentation probably belongs inside the `TypeScript.md` document directly, but I have included it as a separate document for future copy-writers to reference when updating documentation for Fastify version 4. 

There is a bit of information overlap with respect to utilizing TypeBox and json-schema-to-ts which may need to be updated inline with version 4 functionality, so would prefer to defer those edits on `TypeScript.md` for the time being.

File: types/typeProvider.d.ts (hunk 6 above)
Comment (sinclairzx81) [reply types/typeProvider.d.ts]: Renamed to `type-provider.d.ts` as per request.

File: test/type-provider.test.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+  } catch (e) {
+    t.fail()
+  }
+})
+
+test('Should return same instance', t => {
+  t.plan(1)
+  const fastify = Fastify()
+  const next = fastify.withTypeProvider()
+  if (next === fastify) {
```
Comment (zekth) [inline test/type-provider.test.js:20]: This is the node tap way.
```suggestion
  t.equal(next,fastify)
```

**Review by zekth — APPROVED** [review]
Excellent. LGTM

File: test/type-provider.test.js (hunk 7 above)
Comment (L2jLiga) [reply test/type-provider.test.js]: It seems like suggestion was applied and now we have syntax error in this file 😅

File: docs/TypeScript.md (hunk 2 above)
Comment (RafaelGSS) [reply docs/TypeScript.md:332]: I'll create an issue for that when released then.

**Review by RafaelGSS — APPROVED** [review]
LGTM.

**Review by zekth — APPROVED** [review]
Lgtm

**Comment by mcollina** [comment]
Landed! @sinclairzx81 would you like to join as a Fastify collaborator and as a member of @fastify/typescript?

**Comment by sinclairzx81** [comment]
Hey fantastic! :) Am hoping this feature goes on to be well received by the Fastify community and opens the door to many other enhancements in the TypeScript type inference space! :)

@mcollina Thank you so much for the offer to join! I've just send through a quick email on this, so quite happy to take a look into things there. 

Thanks Everyone!
S

File: types/register.d.ts [hunk 8]
```diff
@@ -1,13 +1,14 @@
...
   prefix?: string;
   logLevel?: LogLevel;
   logSerializers?: Record<string, (value: any) => string>;
 }
 
-export type FastifyRegisterOptions<Options> = (RegisterOptions & Options) | (() => RegisterOptions & Options)
+export type FastifyRegisterOptions<Options> = (RegisterOptions & Options) | ((instance: FastifyInstance) => RegisterOptions & Options)
 
 /**
  * FastifyRegister
```
Comment (rickerp) [inline types/register.d.ts:14]: @sinclairzx81  I think I found a missing part in the [FastifyRegister interface](https://github.com/fastify/fastify/pull/3398/files#diff-3ab92153aecdba9f45a29d5fa78392abd148e8e9841436d84e5e712d0542fc94R18-R31) when using the `instance.register` will complain when trying to use a TypeProvider. I believe this is the code missing

```ts
import { RawServerBase } from "./utils";
import { FastifyTypeProvider, FastifyTypeProviderDefault } from "./type-provider";
```
```ts
export interface FastifyRegister<T = void> {
  <Options extends FastifyPluginOptions, Server extends RawServerBase, TypeProvider extends FastifyTypeProvider>(
    plugin: FastifyPluginCallback<Options, Server, TypeProvider>,
    opts?: FastifyRegisterOptions<Options>
  ): T;
  <Options extends FastifyPluginOptions, Server extends RawServerBase, TypeProvider extends FastifyTypeProvider>(
    plugin: FastifyPluginAsync<Options, Server, TypeProvider>,
    opts?: FastifyRegisterOptions<Options>
  ): T;
  <Options extends FastifyPluginOptions, Server extends RawServerBase, TypeProvider extends FastifyTypeProvider>(
    plugin: FastifyPluginCallback<Options, Server, TypeProvider> | FastifyPluginAsync<Options, Server, TypeProvider> | Promise<{ default: FastifyPluginCallback<Options, Server, TypeProvider> }> | Promise<{ default: FastifyPluginAsync<Options, Server, TypeProvider> }>,
    opts?: FastifyRegisterOptions<Options>
  ): T;
}
```

Comment (sinclairzx81) [reply types/register.d.ts:14]: @rickerp Nice! Are you are to test with a couple of community plug ins (like `fastify-swagger`) ? The following are a couple of tests which should help verify the change. (should be able to drop the following in the https://github.com/fastify/fastify/blob/next/test/types/type-provider.test-d.ts file)

```typescript
import fastifySwagger from 'fastify-swagger'

// --------------------------------------------------------------------
// Utility Type (pending object interface refactor)
// --------------------------------------------------------------------

interface FastifyInstanceWithTypeProvider<TypeProvider extends FastifyTypeProvider> extends FastifyInstance<
  RawServerDefault,
  RawRequestDefaultExpression<RawServerDefault>,
  RawReplyDefaultExpression<RawServerDefault>,
  FastifyLoggerInstance,
  TypeProvider
> {}

const fastify = Fastify() // Fastify will use FastifyTypeProviderDefault as default

// --------------------------------------------------------------------
// Local Plugins
// --------------------------------------------------------------------

async function anyTypePlugin(instance: any, options: any) { }

async function typeProviderPlugin(instance: FastifyInstanceWithTypeProvider<TypeBoxProvider>, options: { xyz: string }) { }

// --------------------------------------------------------------------
// Tests
// --------------------------------------------------------------------

// Register should infer for any plugin arguments.
fastify.register(anyTypePlugin, { xyz: '123' }) 

// Register should infer for varying Type Providers (i.e. TypeBoxProvider !== FastifyTypeProviderDefault)
fastify.register(typeProviderPlugin, { xyz: '123' }) 

// Register should infer from existing plugins that use a typed plugin signature (but are non-TypeProvider aware)
fastify.register(fastifySwagger, {}) 
```
If all checks out, you should submit a PR. 

Good work!

Comment (rickerp) [reply types/register.d.ts:14]: No ,I tested wit my own plugins (like "routers")

Comment (RafaelGSS) [reply types/register.d.ts:14]: Could you create an issue to keep tracking this issue?

**Pull request merged**